[#2] 회원가입 기능 추가

hanwix2 opened the pull request:
### #2 회원가입 기능 구현

- [x] MyBatis, MySQL 적용
- [x] 회원가입 기능 (DB에 단순 회원 데이터 insert)
- [x] 회원 데이터 입력값 검증 (유효값, Not Null 등)
- [x] 비밀번호 암호화
- [x] 아이디 중복체크
- [x] 테스트 코드 작성

hanwix2 committed:
[#2] 회원가입 기능 추가

- MyBatis, MySQL 적용
- 단순 회원 데이터 저장

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -0,0 +1,54 @@
+/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
+/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
+/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
+/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;
```
스키마 선언에 필요한 구문만 넣어주세요~

f-lab-dev commented on pom.xml:
```diff
@@ -42,6 +42,17 @@
+        <dependency>
+            <groupId>org.mybatis.spring.boot</groupId>
+            <artifactId>mybatis-spring-boot-starter</artifactId>
+            <version>2.1.4</version>
```
이 버전은 생략이 가능한데 왜 그런걸까요??

hanwix2 replied:
스프링 부트에서는 **`starter `** 라는 라이브러리를 통해 의존성을 쉽게 관리 해줍니다.
자주 사용되는 라이브러리 간의 의존성, 버전 조합을 사전에 정의하여 패키지처럼 제공하여 개발 중 발생하는 의존성 충돌을 방지하는 등 의존성 관리를 용이하게 도와줍니다.

위에서 `spring-boot-starter-parent` 를 정의하고, 각 모듈의 starter를 의존성으로 추가했으므로  
적절한 버전을 자동으로 설정해줍니다.

hanwix2 replied:
다시 한번 확인해보니 mybatis는 의존성이 정의되어 있지 않네요. 

따라서 버전 명시는 필요한 것 같습니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -0,0 +1,24 @@
+
+    private final UserService userService;
+
+    @PostMapping("/users/new")
```
`POST`는 이미 new라는 뜻을 가지고 있습니다~ 이 문서를 참고해보시면 좋을 것 같아요

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -0,0 +1,24 @@
+        int cnt = userService.insertUser(user);
+
+        // return "redirect:/";
+        return cnt + " 회원가입 완료!";
```
리턴을 굳이 해주지 않아도 `200 status code`가 리턴되어 성공했다는 의미를 전달할 수 있습니다. 위에 첨부드린 링크를 읽어보셔도 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/show/showticketingservice/model/User.java:
```diff
@@ -0,0 +1,19 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
이 클래스는 `@Setter`가 꼭 필요할까요? `java immutable` 에 대해 검색해보시면 좋을 것 같아요~

hanwix2 replied:
암호화된 비밀번호를 다시 적용 시키려했지만 불변성에 위배되는군요..

알아보니 Builder 패턴을 사용하면 해결할 수 있을 것 같네요!

f-lab-dev commented on src/main/java/com/show/showticketingservice/repository/UserRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+@Repository
+@RequiredArgsConstructor
+public class UserRepository {
```
`Repository`와 `Mapper`를 따로 두신 이유가 무엇이실까요?

hanwix2 committed:
[#2] Update sql file - 스키마 선언에만 필요한 구문 정리

hanwix2 committed:
[#2] 회원가입 기능 구현

- BCrypt를 이용한 비밀번호 암호화
- ID 중복검사 기능 추가

hanwix2 committed:
[#2] 회원가입 기능 구현 3

- 회원 데이터 입력값 검증(Validation) 코드 추가

f-lab-dev commented on etc/show_ticketing_service.sql:
```diff
@@ -30,25 +7,4 @@ CREATE TABLE `user` (
   `email` varchar(40) NOT NULL,
   `address` varchar(100) DEFAULT NULL,
   PRIMARY KEY (`id`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```
이거는 필요한건데 지워져있네요. 어떤 역할을 하는 구문인지 알아보시면 좋을 것 같네요~

hanwix2 replied:
- **ENGINE:** 데이터베이스 엔진 설정
    - `InnoDB` - MySQL의 기본 엔진으로, 디폴트로 적용되긴 하지만 이 설정이 어느 상황에도 작용한다는 조건이 아니므로 명시하는 것이 좋다.
- **CHARSET:** character set (문자 인코딩 설정)
    - `utf8mb4:` utf-8 인코딩 체계에서 MaxByte 4로 저장 
    - 전세계 모든 언어는 21bit로 3byte 이내로 처리 되어 이모지 같이 4byte로 처리되는 문자열을 처리하기 위해 설정
- **COLLATE:** 문자열 정렬 방식 설정
    - `utf8mb4_0900_ai_ci`
    - 0900: Unicode Collation Algorithm 버전
    - ai: 악센트를 구분 안함
    - ci: 대소문자 구분 안함

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -2,23 +2,43 @@
-    @PostMapping("/users/new")
-    public String insertUser(@RequestBody User user) {
-        int cnt = userService.insertUser(user);
+    @PostMapping("/create_account")
```
`POST` 메소드에는 이미 생성이라는 뜻이 들어있습니다~ 즉 create는 중복표현인데 이 문서를 참고해보시면 좋을 것 같습니다~ 

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

hanwix2 replied:
사실, 실제 사용되는 uri를 찾아보다가 '다음'에서 회원가입을 할 때 create_account를 사용하더라고요, 그래서 참고한 것인데

다른 표현을 한번 찾아보겠습니다!

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -2,23 +2,43 @@
 
-        // return "redirect:/";
-        return cnt + " 회원가입 완료!";
+        if (userService.isIdDuplicated(user.getId()))
```
`중복된 아이디로는 회원가입을 할 수 없다`는 비즈니스 룰입니다. 즉 service layer에 들어가야하는게 아닐까요?

tlsgkr7416 commented on pom.xml:
```diff
@@ -59,6 +59,11 @@
         </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-validation</artifactId>
+        </dependency>
```
test

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/tool/PasswordEncryptor.java:
```diff
@@ -0,0 +1,13 @@
+import org.mindrot.jbcrypt.BCrypt;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class PasswordEncryptor {
+
+    public String encrypt(String password) {
+        return BCrypt.hashpw(password, BCrypt.gensalt());
+    }
+
+}
```
class로 따로 만들어서 메소드를 정의한 이유가 뭔지 궁굼하네요 그리고 @configuration을 작성한 이유도 궁굼합니다.

hanwix2 replied:
일종의 관심사의 분리이죠 ㅎㅎ 암호화 기능은 별도의 클래스에서 담당하는 걸로

그리고 configuration은 component로 하는 것을 잘못 적은것 같네요 ㅠ 체크해주셔서 감사합니다.

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/PasswordEncryptor.java:
```diff
@@ -0,0 +1,13 @@
+import org.mindrot.jbcrypt.BCrypt;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
```
이거는 `@Component`로 정의해야하지 않을까요?

hanwix2 replied:
옙. Component라고 머리는 생각했는데 손은 다른 것을 적었네요 ㅠ 정정하겠습니다!

f-lab-dev commented on src/main/java/com/show/showticketingservice/tool/PasswordEncryptor.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class PasswordEncryptor {
```
이 클래스는 암호화 방식 변경에 따라 확장의 여지가 있을 것 같은데 어떻게 하는게 좋을까요?

hanwix2 committed:
[#2] 회원가입 기능 구현 4

- sql파일 수정(테이블 생성시 부가 설정 명시)
- 암호화 기능 추상화(Encryptor interface 생성)
- ID 중복체크 GET 메소드 구현
- 회원가입 요청 작업중 ID 중복체크 후 가입 로직, service layer로 이동
- POST 메소드 URI 수정
- constructor injection 적용

hanwix2 committed:
[#2] 회원가입 기능 구현 5

- UserController 테스트 코드 작성(UserControllerTest.java)
- User Validation 테스트 코드 작성(UserValidationTest.java)
- id 중복체크, GET method parameter annotation 수정
- User Validation Patter 수정

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -3,44 +3,33 @@
-    // 회원가입 신청 전 추가적인 요청으로 분리 필요.
-    @GetMapping("/check_id")
-    public ResponseEntity<Void> checkIdDuplicated(@RequestParam String id) {
+    @GetMapping("/idCheck/{id}")
```
여기도 고민이 많이 필요해보입니다. `/users/{id}`는 **해당 id에 속한 유저를 가져온다** 라는 행위를 명시하는것인데 그럼 이 URL은 과연 중복체크라는 행위를 잘 표현하고 있을까라는 고민을 계속 하셔야 고민의 깊이가 깊어집니다

hanwix2 replied:
고민해보고 수정하겠습니다!

f-lab-dev reviewed (commented):
커밋메세지를 보면 1,2,3,4,5 이렇게 숫자를 쓰고계신데 이런 숫자는 의미를 파악하기 힘들기 때문에 지양해야합니다~
굳이 쓰시지 않아도 될 것 같아요

hanwix2 committed:
[#2] ID 중복 체크 GET method url 수정

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -24,8 +24,8 @@
 
-    @GetMapping("/idCheck")
-    public ResponseEntity<Void> checkIdDuplicated(@RequestParam String id) {
+    @GetMapping("/{id}/exists")
```
기존것보다 이게 더 깔끔해보이네요~ 👍 
객체지향이나 이런 URL 구조같은 설계는 계속 고민을 하셔야 고민의 깊이가 깊어집니다.
앞으로도 계속 여러 고민들을 해보시면 좋을 것 같습니다

f-lab-dev commented on src/test/java/com/show/showticketingservice/controller/UserControllerTest.java:
```diff
@@ -72,8 +72,7 @@ public void duplicatedIdSignUp() throws Exception {
     public void idNotDuplicated() throws Exception {
-        mvc.perform(get("/users/idCheck")
-                .param("id", "newId1"))
+        mvc.perform(get("/users/newId1/exists"))
```
보통 `newId1`가 아닌 숫자가 들어가지 않나요?? `id`와 `userId`는 구분되는게 좋을 것 같아요~

hanwix2 replied:
`newId1`란 아이디는 잘못 지은 것 같긴하네요.

실제 서비스에서는 아이디를 문자와 숫자의 조합으로 입력할텐데, 테스트 코드에서는 숫자만 적어도 될까요? 
전에 보내주신 API디자인 지침에서는 예시가 숫자이긴 했지만 약간 혼동되네요.

hanwix2 replied:
id와 userId를 구분하라는 것은 id를 DB에 저장되는 회원을 구분하는 수로 사용하라는 말씀 같은데, 맞나요?

그럼 userId의 중복 검사에 해당하는 기능은 url을 다른 식으로 짜야하는 것일까요?

hanwix2 committed:
[#2] UserId 중복 체크 GET method url 수정

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/UserController.java:
```diff
@@ -0,0 +1,34 @@
+        return HttpStatusResponse.HTTP_STATUS_CONFLICT;
+    }
+
+    @GetMapping("/user-exists/{userId}")
```
이거는 어떤게 더 직관적이고 좋은 표현일지 계속 고민을 해보시면 좋을 것 같네요~

hanwix2 committed:
[#4] 로그인 기능 구현

- 로그인 기능 구현
- Data Access를 위한 Mapper 인터페이스만 사용 (Repository 삭제)

hanwix2 committed:
[#4] 로그인 기능 관련 수정

- DB user table 수정: id와 userId 구분 & userType 추가
- user table 수정에 따른 관련 코드(테스트 포함) 수정
- User 모델을 UserRequest와 UserResponse로 구분
- userId 중복 검사에 대한 예외 처리 추가
- 예외처리 중 로거에 보내는 메세지 수정
- UserService의 getUser메소드에서 sql쿼리를 최소화 하기 위한 로직 수정
- PasswordEncryptor 클래스명 BCryctPasswordEncryptor로 변경

hanwix2 committed:
[#4] 중복 로그인 예외처리

이미 로그인 된 상태에서 중복으로 로그인 시도 시 예외처리

hanwix2 committed:
[#4] UserType 속성 Enum으로 처리 및 피드백 반영

Enumeration 관련
- User의 UserType 속성 Enumeration화
- TypeHandler로 Enum 처리

피드백 관련
- UserConstant.java 생성: session attribute key값 상수 처리
- PasswordEncryptor 클래스, 변수 이름 수정
- @ResponseStatus 어노테이션 사용
- id 중복확인 sql 쿼리문 수정: boolean 형 반환

hanwix2 committed:
[#4] UserMapper sql 쿼리문 수정

getUserByUserId sql쿼리문 수정 - SELECT 하는 컬럼명 명시

hanwix2 committed:
Merge pull request #13 from f-lab-edu/feature/4/jh

[#4] 로그인 기능 구현
